### ISSUE TITLE
"Setting up TensorFlow for Development" no longer works with recent TF versions

### ISSUE THREAD

**Issue by rdadolf** [tensorflow/tensorflow#1704]
### Environment info
Operating System: OS X 10.11, Ubuntu 15.10
Source build, commit hash: e39d8feebb9666a331345cd8d960f5ade4652bba

### Steps to reproduce
I have TF as a submodule and build it local to the project. I use more or less the exact same commands in the [development setup](https://www.tensorflow.org/versions/r0.7/get_started/os_setup.html#setting-up-tensorflow-for-development). The only minor change is that I don't run `python setup.py develop` (under OS X, this breaks my Anaconda python installation completely via screwing with the site-packages). So the steps are (in Makefile syntax):

```bash
git submodule update --init --recursive
cd $(TF_DIR) && ./configure
cd $(TF_DIR) && bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
cd $(TF_INSTALL_DIR) && ln -s $(addprefix ../../,$(wildcard $(TF_DIR)/bazel-bin/tensorflow/tools/pip_package/build_pip_package.runfiles/*)) .
$(TF_INSTALL_DIR) && ln -s $(addprefix ../../,$(wildcard $(TF_DIR)/tensorflow/tools/pip_package/*)) .
```

The prefix commands are for making globs work in Make. I have confirmed they expand to the same symlink commands.

### Issue
Importing tensorflow no longer works using development guidelines. Specifically, assuming tensorflow is cloned into $TF_DIR, symlinked into `$TF_INSTALL_DIR`, and built by bazel in $BAZEL_TMP, this no longer works: (output sanitized manually)

```
$ PYTHONPATH=$TF_INSTALL_DIR python
Python 2.7.10 (default, Oct 14 2015, 16:09:02) 
[GCC 5.2.1 20151010] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "$TF_INSTALL_DIR/tensorflow/__init__.py", line 23, in <module>
    from tensorflow.python import *
  File "$TF_INSTALL_DIR/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.core.framework.graph_pb2 import *
  File "$TF_INSTALL_DIR/tensorflow/core/framework/graph_pb2.py", line 6, in <module>
    from google.protobuf import descriptor as _descriptor
  File "$BAZEL_TMP/tensorflow/bazel-out/local_linux-opt/bin/tensorflow/tools/pip_package/build_pip_package.runfiles/google/protobuf/descriptor.py", line 46, in <module>
    from google.protobuf.pyext import _message
  File "$BAZEL_TMP/tensorflow/bazel-out/local_linux-opt/bin/tensorflow/tools/pip_package/build_pip_package.runfiles/google/protobuf/pyext/__init__.py", line 2, in <module>
    __import__('pkg_resources').declare_namespace(__name__)
  File "/usr/local/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2226, in declare_namespace
    _handle_ns(packageName, path_item)
  File "/usr/local/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2195, in _handle_ns
    path.sort(key=sort_key)
  File "/usr/local/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 2193, in sort_key
    return sys_path.index(_normalize_cached(os.sep.join(parts)))
ValueError: '$BAZEL_TMP/tensorflow/bazel-out/local_linux-opt/bin/tensorflow/tools/pip_package/build_pip_package.runfiles' is not in list
```

Linux output shown, but OS X also fails with a similar exception.

### What have you tried?

I believe the issue comes down to the same one as #1535. I am creating this issue separately so that if it turns out to be a different problem, it isn't muddled. If it's the same, we can close it.

The exception is a result of namespace package resolution in google.protobuf trying to find `$BAZEL_TMP/tensorflow/bazel-out/local_linux-opt/bin/tensorflow/tools/pip_package/build_pip_package.runfiles/google/protobuf/pyext`. It is trying to find it in a list of system paths which includes `$TF_INSTALL_DIR`. Recall that due to the development setup rules, the files in `$TF_INSTALL_DIR` are symlinks to files in `$BAZEL_TMP`. Specifically, `$TF_INSTALL_DIR/google/protobuf` is the same directory as `$BAZEL_TMP/google/protobuf`. (I have verified this with `pwd -P`.

What seems to be going on is that Python is attempting to unify the protobuf namespace package, but its filename (`$BAZEL_TMP/google/protobuf/pyext`) does not match the `sys.path` entry (`$TF_INSTALL_DIR`).

**Note:** This only appears to be a problem when trying to use the development install rules, and it only appears to be a problem when there's another tensorflow or protobuf installed in the system somewhere else. At least on OS X; the other version of tensorflow on my Linux machine is in use by other users, so I can't uninstall it safely. I have not attempted to create a clean virtual environment just for my development directory (for one, because virtual environments are a pain in the neck; and two, that's part of the whole point of having a local build of tensorflow).

Workaround: Replacing the symlinks with a recursive copy (from `$BAZEL_TMP` to `$TF_INSTALL_DIR`) is apparently sufficient to convince Python that it's unified the namespace packages correctly, and it chooses the `$TF_INSTALL_DIR` specified in `$PYTHONPATH`.

So I don't know what the appropriate solution here is, but the problem seems to be the result of a convergence of semi-hacky solutions, any one of which alone might not be an issue, but work together to cause problems:

 - There's no clean way to extract the necessary files from a built tensorflow without going through pip install. Perhaps there's a method using pip I'm not aware of (maybe something with `--target`?).
 - Python namespace packages have poor path resolution semantics. It looks like this ship has sailed.
 - The symlink approach is kind of coarse and tricky to script around.

**Comment by wangg12** [comment]
Hi @rdadolf , I met this problem just now by coincidence. I tried rebuild the project but the problem was   not solved. Then I tried `pip uninstall tensorflow` and `python setup.py develop`. It was fixed.

**Comment by rdadolf** [comment]
Yep, this is one workaround. I mentioned that this did alleviate the issue in the "**Note:**" piece above, and I think a similar approach was mentioned in the other issue I referenced. I should have probably placed it more prominently.

This doesn't help a whole lot with shared systems, however. I have other folks who work on an older, stable install of TF on the same system (and sometimes I'm one of them), so uninstalling isn't a solution in my case. I believe virtualenv's have been tried, and they may work here, but one of the whole ideas behind a local install was to avoid that route.

I'm glad that this solved your problem, though.

To devs: if it looks like this problem is too much of a pain for not enough benefit, another stopgap would be to update the development setup docs to point out that you either need to have protobuf version compatibility and/or uninstall other versions of tensorflow to do this. Might save some folks in @wangg12's situation some headache.

**Comment by wangg12** [comment]
@rdadolf I'm also using a shared system. But for python, I installed an anaconda just for me. So there won't be any problem to influence others' environment.

**Comment by orionr** [comment]
For those that land on this ticket from Google searches, you might also want to check out https://github.com/tensorflow/tensorflow/issues/2591 which outlines a typo in the `master` docs.

**Comment by rdadolf** [comment]
Just a followup on @orionr's mention: the typo in #2591 is a recent addition: prior to around May 24 2016, the line in the setup documentation didn't have this issue, since the `org_tensorflow` part didn't exist. (This is the [relevant diff](https://github.com/tensorflow/tensorflow/pull/2336/files#diff-1cf444a338533a8469b274d1cd1b7229).) So the original issue in this thread is not a symptom.

Thank you for mentioning it, though. They are similar issues, and it's tricky to find all of the relevant issue/commit threads in github. This helped me eliminate another recent problem regarding bazel versioning.

**Comment by rdadolf** [comment]
Given the [recent flux](https://github.com/tensorflow/tensorflow/commit/14ac2235699509f512b44b71160239c153ab413d), I thought I'd update this issue based on a more recent version (79174af).

First, I've updated to bazel 2.3 to avoid [this problem](https://github.com/tensorflow/tensorflow/issues/2497#issuecomment-221631692), so the relevant commands now use the `org_tensorflow` syntax (with the typo fix).

The changes to protobuf have not solved these issues, they've just broken them in different ways.

**Change in behavior regarding manual import**

It looks like the change has broken the way I was importing TensorFlow in the original issue. I.e, running through the TensorFlow development setup instructions without the `setup.py` step and just importing from the directory no longer works. The specific commands are (*not* in Makefile syntax this time, to make things easier to read):

```
cd $TF_DIR && ./configure
cd $TF_DIR && bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
rm -r $TF_INSTALL_DIR
mkdir $TF_INSTALL_DIR
cd $TF_INSTALL_DIR && cp -r $TF_DIR/bazel-bin/tensorflow/tools/pip_package/build_pip_package.runfiles/org_tensorflow/* .
cd $TF_INSTALL_DIR && cp -r $TF_DIR/tensorflow/tools/pip_package/* .
```

Example 1: In a clean environment, `protobuf` is not found.
```
PYTHONPATH=$TF_INSTALL_DIR python
>>> import tensorflow as tf
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "$TF_INSTALL_DIR/tensorflow/__init__.py", line 23, in <module>
    from tensorflow.python import *
  File "$TF_INSTALL_DIR/tensorflow/python/__init__.py", line 58, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "$TF_INSTALL_DIR/tensorflow/python/__init__.py", line 52, in <module>
    from tensorflow.core.framework.graph_pb2 import *
  File "$TF_INSTALL_DIR/tensorflow/core/framework/graph_pb2.py", line 6, in <module>
    from google.protobuf import descriptor as _descriptor
ImportError: No module named google.protobuf

Error importing tensorflow.  Unless you are using bazel,
you should not try to import tensorflow from its source directory;
please exit the tensorflow source tree, and relaunch your python interpreter
from there.
```

Example 2: In a mixed environment (i.e., one with a system-installed version of tensorflow and a custom verstion. In this case, they are the *same* version of protobuf, so we don't get an error as in the original issue):
```
PYTHONPATH=$TF_INSTALL_DIR python
>>> import tensorflow as tf
>>> tf.__file__
$TF_INSTALL_DIR/tensorflow/__init__.pyc
>>> import google.protobuf
>>> google.protobuf.__file__
$SYSTEM_SITE_PACKAGES/google/protobuf/__init__.pyc
```

It looks like the changes have made importing directly impossible. That's unfortunate, but I understand that it wasn't a supported behavior in the first place. So I bit the bullet and tried to make things work as in the first bullet point I mentioned in the original file (working through `setup.py` or `pip`).

**Installing via "Setting up TensorFlow for Development"**

Same commands as before, but this time using the `setup.py` as written (and a separate build directory):
```
cd $TF_DIR && ./configure
cd $TF_DIR && bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
rm -r $TF_BUILD_DIR
mkdir $TF_BUILD_DIR
cd $TF_BUILD_DIR && ln -s $TF_DIR/bazel-bin/tensorflow/tools/pip_package/build_pip_package.runfiles/org_tensorflow/* .
cd $TF_BUILD_DIR && ln -s $TF_DIR/tensorflow/tools/pip_package/* .
python $TF_BUILD_DIR/setup.py develop
```

Note: this is *not* safe. This can overwrite files in your `site-packages` directory, including replacing existing `easy_install` and `wheel` scripts and modifying your existing `sys.path` through `.pth` files. If there's a flaw here (or, say a mismatch between the version of `setuptools` that TensorFlow just installed and your other tools like, say `pip` or `conda`) and you're not using some type of sandboxing (virtualenv, VM, etc.), python is basically broken until you manually remove the offending files TensorFlow installs (and you won't be able to use the `--uninstall` option because of the same underlying issue). Also, if you already had a working version of tensorflow, it's now gone, as they conflict and `setup.py` will remove the other. So as before, I hope you were working in a sandbox.

So this sort of works, but it doesn't solve my original problem of having any sort of control over where the library was installed. So there are a couple of possible solutions:

**Flags to `setup.py`**
`setup.py develop` takes an `--install-dir` flag. Long story short, it doesn't work. 1) it requires pre-modifying your environment, 2) it doesn't *actually* obey the flag---it still modifies files in your default `site-packages`, and 3) it doesn't actually work: python doesn't find the library, even when the directory is specified in `PYTHONPATH`.

Counter-intuitively, `setup.py install` has different options, including a `--prefix` flag, which doesn't do quite the same thing. Except it turns out that this the same problems.

Ultimately, both of these use `.pth` files as an underlying mechanism. Because python only reads `.pth` files from four specific directories, these methods are unworkable.

**Flags to pip**
So the remaining option is to forget the "development" options altogether and attempt to use the pip package. Recent versions of `pip` take a `--prefix` option, which is supposed to support this. So the commands are:

```
cd $TF_DIR && ./configure
cd $TF_DIR && bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
$TF_DIR/bazel-bin/tensorflow/tools/pip_package/build_pip_package $TF_BUILD_DIR
pip install $TF_BUILD_DIR/tensorflow-0.8.0-py2-none-any.whl --prefix=$TF_INSTALL_DIR
```
and we get this output:
```
Thu Jun 2 16:27:58 EDT 2016 : === Using tmpdir: /var/folders/mw/kf4sryxx1xvgh2kyg9_yxvsw0000gn/T/tmp.XXXXXXXXXX.PYkoZMt7
/var/folders/mw/kf4sryxx1xvgh2kyg9_yxvsw0000gn/T/tmp.XXXXXXXXXX.PYkoZMt7 ~/scratch/tensorflow
Thu Jun 2 16:27:58 EDT 2016 : === Building wheel
~/scratch/tensorflow
Thu Jun 2 16:28:10 EDT 2016 : === Output wheel file is in: $TF_BUILD_DIR
Processing $TF_BUILD_DIR/tensorflow-0.8.0-py2-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): six>=1.10.0 in $SYSTEM_SITE_PACKAGES (from tensorflow==0.8.0)
Collecting protobuf==3.0.0b2 (from tensorflow==0.8.0)
  Using cached protobuf-3.0.0b2-py2.py3-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): numpy>=1.10.1 in $SYSTEM_SITE_PACKAGES (from tensorflow==0.8.0)
Requirement already satisfied (use --upgrade to upgrade): wheel in $SYSTEM_SITE_PACKAGES (from tensorflow==0.8.0)
Requirement already satisfied (use --upgrade to upgrade): setuptools in $SYSTEM_SITE_PACKAGES/setuptools-21.2.1-py2.7.egg (from protobuf==3.0.0b2->tensorflow==0.8.0)
Installing collected packages: protobuf, tensorflow
Successfully installed protobuf tensorflow
```

And we try importing...
```
PYTHONPATH=$TF_INSTALL_DIR/lib/python2.7/site-packages python
>>> import tensorflow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "$TF_INSTALL_DIR/lib/python2.7/site-packages/tensorflow/__init__.py", line 23, in <module>
    from tensorflow.python import *
  File "$TF_INSTALL_DIR/lib/python2.7/site-packages/tensorflow/python/__init__.py", line 58, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "$TF_INSTALL_DIR/lib/python2.7/site-packages/tensorflow/python/__init__.py", line 52, in <module>
    from tensorflow.core.framework.graph_pb2 import *
  File "$TF_INSTALL_DIR/lib/python2.7/site-packages/tensorflow/core/framework/graph_pb2.py", line 6, in <module>
    from google.protobuf import descriptor as _descriptor
ImportError: No module named google.protobuf

Error importing tensorflow.  Unless you are using bazel,
you should not try to import tensorflow from its source directory;
please exit the tensorflow source tree, and relaunch your python interpreter
from there.
```

Which brings us back to a similar problem. It seems nigh-on impossible to get tensorflow to import protobuf using the new system unless it's installed in the default user or system directory. I believe this boils down to the same namespace packaging issue, but that's just a conjecture at this point.

**TL;DR**
The new protobuf mechanism doesn't actually solve this problem. Installing tensorflow in a non-standard location seems to cause protobuf imports to fail regardless of how or where.

**Comment by orionr** [comment]
@rdadolf - thanks for updating to the latest, which eliminates the issue I was commenting on. This might be silly, but have you considered adjusting your `PYTHONPATH` variable? With that you might be able to get it to find your tensorflow package as well as protobuf. Basically something like this, if `import google.protobuf` succeeded when you ran `python` without the path:

```
PYTHONPATH=$TF_INSTALL_DIR/lib/python2.7/site-packages:$PYTHONPATH python
>>> import tensorflow
...
```
And, of course, using virtualenv might help (it works great for me), but it looks like you specifically don't want to go that route. Good luck!

**Comment by rdadolf** [comment]
Yep. That's the ultimate goal here: place tensorflow in directory X, and point `PYTHONPATH` to it, and write `import tensorflow`, where X is a directory of my choosing. Each one of these examples actually sets `PYTHONPATH` to `TF_INSTALL_DIR` before running, just like you mention. It certainly *seemed* like it should be that easy. Unfortunately, it seems that either due some combination of the packaging tools and the way the TensorFlow and/or protobuf packages are written, it's not.

I do understand the virtualenv perspective. Without getting into too much of a holy war, virtualenv takes a very different approach to solving the problem, one that is pretty narrow in scope. Part of my hope was that I would only have to have one solution for dealing with versioning, environment, and consistency. virtualenv can only handle a small piece of the python-only part of that. For instance, there's a collection of training data and preprocessing which needs similar solutions: externally managed, place it in directory X, point Y to it, etc. Virtualenv isn't much help there. Same with some low-level instrumentation code which we use. And this is compounded by running on a variety of machines and platforms. So while I'll admit I'm a little biased because I don't much like virtualenv's approach to the world, I'd also really like to avoid adding more complexity in a lot of places just because of idiosyncrasies in the way tensorflow's package is written. (Especially since it was working two weeks ago!)

**Comment by orionr** [comment]
Only thing I would suggest is to concatenate to the existing PYTHONPATH rather than overriding it. Take one more look at my previous comment and the :$PYTHONPATH inserted in the middle. If that doesn't work I guess we'll need to hear from the real maintainers.

**Comment by yaroslavvb** [comment]
This would be nice to figure out with "just use PYTHONPATH" way.  This way one could start `ipython notebook` to play around with their tensorflow changes in Jupyter, but current way is limiting you to `simple_console.py`

**Comment by keveman** [comment]
These set of commands work for me :

    mkdir tensorflow
    export TF_INSTALL_DIR=/tmp/tensorflow
    pip install --prefix=$TF_INSTALL_DIR https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.9.0rc0-cp27-none-linux_x86_64.whl
    Collecting tensorflow==0.9.0rc0 from https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.9.0rc0-cp27-none-linux_x86_64.whl
    Using cached https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.9.0rc0-cp27-none-linux_x86_64.whl
    Requirement already satisfied (use --upgrade to upgrade): six>=1.10.0 in /usr/local/lib/python2.7/dist-packages (from tensorflow==0.9.0rc0)
    Requirement already satisfied (use --upgrade to upgrade): protobuf==3.0.0b2 in /usr/local/lib/python2.7/dist-packages (from tensorflow==0.9.0rc0)
    Requirement already satisfied (use --upgrade to upgrade): wheel in /usr/lib/python2.7/dist-packages (from tensorflow==0.9.0rc0)
    Requirement already satisfied (use --upgrade to upgrade): numpy>=1.8.2 in /usr/lib/python2.7/dist-packages (from tensorflow==0.9.0rc0)
    Requirement already satisfied (use --upgrade to upgrade): setuptools in /usr/local/lib/python2.7/dist-packages (from protobuf==3.0.0b2->tensorflow==0.9.0rc0)
    Installing collected packages: tensorflow
    Successfully installed tensorflow

    PYTHONPATH=$TF_INSTALL_DIR/lib/python2.7/site-packages:$PYTHONPATH python
    Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
    [GCC 4.8.2] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import tensorflow as tf
    tf>>> tf.__file__
    '/tmp/tensorflow/lib/python2.7/site-packages/tensorflow/__init__.pyc'

    mkdir /tmp/tensorflow
    export TF_INSTALL_DIR=/tmp/tensorflow
    
I am on a Ubuntu 14.04, though. Can you give this a shot? Having said that, `virtualenv` is our tested installation methodology. We need lot more people asking for this way on installation before we can start thinking about supporting and testing this. Feel free to keep this open in case you really want to see this supported, otherwise please close the issue.

**Comment by keveman** [comment]
@rdadolf Closing this one out due to inactivity. I don't see anything fundamentally broken about using `pip` with `--prefix` and using `$PYTHONPATH` later, at least in the platforms I tried. Please feel free to reopen if you think this is still an issue worth solving.
